lib/fastlane/plugin/appcenter/helper/appcenter_helper.rb

```diff
@@ -229,50 +234,119 @@ def self.upload_symbol(api_token, owner_name, app_name, symbol, symbol_type, sym
+
+        case response.status
+        when 200...300
+          chunk_size = response.body['chunk_size']
+          unless chunk_size.is_a? Integer
+            UI.error("Set metadata didn't return chunk size: #{response.status}: #{response.body}")
+            false
+          else
+            UI.message("Metadata set")
+            chunk_size
+          end
+        when 401
+          UI.user_error!("Auth Error, provided invalid token")
+          false
+        else
+          UI.error("Error setting metadata: #{response.status}: #{response.body}")
+          false
+        end
+      end
+
+      # Verifies a successful upload to App Center
+      # returns:
+      # successful upload response body.
+      def self.finish(finish_url, api_token, owner_name, app_name, upload_id, timeout)
+        connection = self.connection(finish_url)
 
+        UI.message("DEBUG: POST #{finish_url}") if ENV['DEBUG']
+        UI.message("DEBUG: POST body <data>\n") if ENV['DEBUG']
+        response = connection.post do |req|
+          req.options.timeout = timeout
+          req.headers['internal-request-source'] = "fastlane"
+        end
         UI.message("DEBUG: #{response.status} #{JSON.pretty_generate(response.body)}\n") if ENV['DEBUG']
 
         case response.status
         when 200...300
-          UI.message("Binary uploaded")
-          self.update_release_upload(api_token, owner_name, app_name, upload_id, 'committed')
+          if response.body['error'] == false
+            UI.message("Upload finished")
+            self.update_release_upload(api_token, owner_name, app_name, upload_id, 'uploadFinished')
+          else
+            UI.error("Error finishing upload: #{response.body['message']}")
+            false
+          end
         when 401
           UI.user_error!("Auth Error, provided invalid token")
           false
         else
-          UI.error("Error uploading binary #{response.status}: #{response.body}")
-          self.update_release_upload(api_token, owner_name, app_name, upload_id, 'aborted')
-          UI.error("Release aborted")
+          UI.error("Error finishing upload: #{response.status}: #{response.body}")
           false
         end
       end
 
+      # upload binary for specified upload_url
+      # if succeed, then commits the release
+      # otherwise aborts
+      def self.upload_build(api_token, owner_name, app_name, file, upload_id, upload_url, content_type, chunk_size, timeout)
+        block_number = 1
+
+        File.open(file).each_chunk(chunk_size) do |chunk|
+          upload_chunk_url = "#{upload_url}&block_number=#{block_number}"
+          connection = self.connection(upload_chunk_url, true)
+
+          UI.message("DEBUG: POST #{upload_chunk_url}") if ENV['DEBUG']
+          UI.message("DEBUG: POST body <data>\n") if ENV['DEBUG']
+          response = connection.post do |req|
+            req.options.timeout = timeout
```

Comment:
This changes the definition of timeout, since we now have `timeout * chunk_count` in total.

I see two use cases for timeout, before this PR:
- Increase it so that large uploads don't fail.
- Reduce it to avoid getting stuck in a user's CI pipeline, when App Center is slow to accept uploads for whatever reason.

I think keeping `timeout` the way you've done it here is fine, but we should be clear that it's different from what we used to have. We already upload symbols (and `timeout` does not work for them, unfortunately) as well as builds, and anecdotal evidence points to the first use case above. But I think this warrants a note in the documentation & README.

Reply:
There are also many steps involved in creating a release, which translates into several API calls before uploading the chunk itself, so I don't think it was correct even before.

We could make the timeout more global and remember time left for the entire process.

Reply:
Added a note on request timeout in the readme.